readthedocs/doc_builder/builder.py

```diff
@@ -0,0 +1,443 @@
+import os
+from collections import defaultdict
+
+import structlog
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
+
+from readthedocs.builds.constants import EXTERNAL
+from readthedocs.doc_builder.config import load_yaml_config
+from readthedocs.doc_builder.loader import get_builder_class
+from readthedocs.doc_builder.python_environments import Conda, Virtualenv
+from readthedocs.projects.exceptions import RepositoryError
+from readthedocs.projects.models import Feature
+from readthedocs.projects.signals import after_build, before_build, before_vcs
+
+log = structlog.get_logger(__name__)
+
+
+class DocumentationBuilder:
+    def __init__(
+        self,
+        data,  # Celery task's ``data`` object
+    ):
+        """Initializer.
+
+        :param data: object with all the data grabbed by Celery task in
+        ``before_start`` and used as a way to share data with this class
+        by-directionally.
+
+        :type data: readthedocs.projects.tasks.builds.TaskData
+
+        """
+        self.data = data
+
+    def vcs(self):
+        """
+        Perform all VCS related steps.
+
+        1. clone the repository
+        2. checkout specific commit/identifier
+        3. load the config file
+        4. checkout submodules
+        """
+        # Make dirs if it does not exist to clone the repository under it
+        if not os.path.exists(self.data.project.doc_path):
+            os.makedirs(self.data.project.doc_path)
+
+        if not self.data.project.vcs_class():
+            raise RepositoryError(
+                _('Repository type "{repo_type}" unknown').format(
+                    repo_type=self.data.project.repo_type,
+                ),
+            )
+
+        environment = self.data.environment_class(
+            project=self.data.project,
+            version=self.data.version,
+            build=self.data.build,
+            environment=self.get_vcs_env_vars(),
+            # Force the ``container_image`` to use one that has the latest
+            # ca-certificate package which is compatible with Lets Encrypt
+            container_image=settings.RTD_DOCKER_BUILD_SETTINGS["os"]["ubuntu-20.04"],
+        )
+        with environment:
+            before_vcs.send(
```

Comment:
Should this be sent in the `pre_checkout` function?

Reply:
Good point. I think you are right. However, as `pre_checkout` does not really work now it won't make any difference, but I think it makes sense to call `before_vcs` there for the future.